### PR TITLE
Rook - Piece specific move logic code

### PR DESCRIPTION
**Pull request by brantwellman**
- Rook model logic
  - is_move_allowed? method completed
  - legit_moves method completed
- Test
  - tests in piece_test.rb

### CODE REVIEW

**Comment by kenmazaika** [comment]
Pull requests during office hours when I'm rambling on.  Haha :)

**Comment by brantwellman** [comment]
I was sooo close! I had to get it in before the meeting ended!  :D

**Comment by bobbyali** [comment]
I don't understand what the legit_moves code is doing. Might be because it's midnight here any my brain has shut off for the night, so I'll take another look tomorrow. One thought though - the variable names (tiles, moves1-4 etc) could be clearer, and that might make it easier to follow?

**Comment by bobbyali** [comment]
Ah I see, tiles is referring to the number of tiles that you're proposing moving the rook away from its current position. (So you start one tile away, then gradually move out further.) OK, I think the code makes sense now.

**Comment by brantwellman** [comment]
That's exactly it. I had originally called that array squares, but Nan and I worked together today a bit and she was calling it tiles. So for the sake of trying to be consistent, I change the name.

File: app/models/rook.rb [hunk 1]
```diff
@@ -2,16 +2,52 @@ class Rook < Piece
...
+		move_logic_is_valid = false
+		if legit_moves.include?	[new_x, new_y]
+			move_logic_is_valid = true
+		end
+		# is there an obstruction in the way? the game model
+		if move_logic_is_valid
+		# if is_move_obstructed returns true, then return false for method
+			return !self.game.is_move_obstructed?(self.id, new_x, new_y)
+		else
+			return nil
```
Comment (kenmazaika) [inline app/models/rook.rb]: You should return `true` / `false` from methods that are named with a question mark at the end.

File: app/models/rook.rb [hunk 2]
```diff
@@ -2,16 +2,52 @@ class Rook < Piece
...
+			# Up move
+			move3 = [x_init, y_move1]
+			# Down move
+			move4 = [x_init, y_move2]
+
+			moves.push(move1, move2, move3, move4)
+
+		end
+
+		moves.delete_if {|moves| moves.include?(nil)}
```
Comment (kenmazaika) [inline app/models/rook.rb]: I think:

```
moves.flatten!
```

Might do what you're doing here in a simpler way.

File: test/models/piece_test.rb [hunk 3]
```diff
@@ -75,4 +75,22 @@ class PieceTest < ActiveSupport::TestCase
 
 	end
 
+	test "check rook moves are legit" do
+		piece = FactoryGirl.create(:rook, :x_coord => 0, :y_coord => 0, :color => "white")
```
Comment (kenmazaika) [inline test/models/piece_test.rb]: You should add tests for obstructed cases.

**Comment by kenmazaika** [comment]
Awesome work, Brant!  Impressive! :)

File: test/models/piece_test.rb (hunk 3 above)
Comment (brantwellman) [inline test/models/piece_test.rb]: The is_move_obstructed? code, which in the game model, is already being tested in game_test.rb? Is it necessary to test that logic again for each individual piece-type?

Comment (kenmazaika) [inline test/models/piece_test.rb]: I think you should make a test to make sure:

https://github.com/theFirehoseProject/chess/pull/36/files#diff-c71ea485fb67e0e0fbb7e3b5170a787dR13

That line is actually doing what you expect.  For example, right now I could replace that line with:

```
return true
```

and all the tests would pass.

Comment (brantwellman) [inline test/models/piece_test.rb]: got it!

Comment (brantwellman) [inline test/models/piece_test.rb]: Test added!

File: test/models/piece_test.rb [hunk 4]
```diff
@@ -75,4 +75,26 @@ class PieceTest < ActiveSupport::TestCase
...
+
+		actual = piece.is_move_allowed?(0, 8)
+		assert_not actual
+
+		actual = piece.is_move_allowed?(8, 0)
+		assert_not actual
+
+		piece_obstruction = FactoryGirl.create(:pawn, :x_coord => 0, :y_coord => 1, :color => "white")
+		actual = piece.is_move_allowed?(0, 2)
+		assert actual, "There should be an obstruction preventing the piece to move"
```
Comment (kenmazaika) [inline test/models/piece_test.rb]: shouldn't we assert that the move is not allowed:

```
assert ! actual
```

**Comment by brantwellman** [comment]
I'm going to be merging this branch into the Move-Queen branch because of a bunch of additional changes I made in that one. The Rook and Queen code will be in that branch when I push up more commits.

**Pull request closed without merging**